scripts/usb_reset.py

```diff
@@ -14,46 +14,266 @@
+                        help="restore ownership to uid gid")
+
+    cleanup = subparsers.add_parser("cleanup", help="clean up chroot directory")
+    cleanup.add_argument("-d", dest="dom_id", type=int, required=True,
+                         help="specify the dom_id of the VM")
+
+    return parser.parse_args()
 
-#USBDEVFS_RESET _IO('U', 20)
-USBDEVFS_RESET = (ord('U') << 8) | 20
 
+def get_root_dir(dom_id):
+    return "/var/xen/qemu/{}/".format(dom_id)
 
+
+def get_cg_dir(dom_id):
+    return "/sys/fs/cgroup/devices/qemu-{}/".format(dom_id)
+
+
+# throw IOError, ValueError
 def read_int(path):
     with open(path) as f:
         return int(f.readline())
 
 
-def reset_device(bus, dev):
-    with open("/dev/bus/usb/{0:03d}/{1:03d}".format(bus, dev), "w") as f:
-        fcntl.ioctl(f.fileno(), USBDEVFS_RESET, 0)
+def dev_path(device):
+    pat = re.compile(r"^\d+-\d+(\.\d+)*$")
+    if pat.match(device) is None:
+        log.debug("Unexpected device node: {}".format(device))
+        exit(1)
+    try:
+        bus = read_int("/sys/bus/usb/devices/{}/busnum".format(device))
+        dev = read_int("/sys/bus/usb/devices/{}/devnum".format(device))
+        return "/dev/bus/usb/{0:03d}/{1:03d}".format(bus, dev)
+    except (IOError, ValueError) as e:
+        log.debug("Failed to get device path {}: {}".format(device, str(e)))
+        exit(1)
 
 
-if __name__ == "__main__":
-    log.logToSyslog(level=logging.DEBUG)
+def get_ctl(path, mode):
+    try:
+        st = os.stat(path)
+    except OSError:
+        return None
+    t = ""
+    if S_ISBLK(st.st_mode):
+        t = "b "
+    elif S_ISCHR(st.st_mode):
+        t = "c "
+    if t and (mode == "r" or mode == "rw"):
+        return t + str(os.major(st.st_rdev)) + ":" + str(os.minor(
+            st.st_rdev)) + " " + mode
+    return None
+
+
+def device_ctl(path, dom_id, allow):
+    cg_dir = get_cg_dir(dom_id)
+    file_name = "devices.allow" if allow else "devices.deny"
+    try:
+        with open(cg_dir + file_name, "w", 0) as f:
+            ctl = get_ctl(path, "rw")
+            if ctl:
+                f.write(ctl)
+    except IOError as e:
+        # log and continue
+        log.debug("Failed to {} {}: {}".format(
+            "allow" if allow else "deny", path, str(e)))
+
+
+def setup_cgroup(dom_id, pid):
+    cg_dir = get_cg_dir(dom_id)
+
+    try:
+        os.mkdir(cg_dir, 0755)
+    except OSError as e:
+        if e.errno != errno.EEXIST:
+            log.debug("Failed to create cgroup: {}".format(cg_dir))
+            exit(1)
+
+    try:
+        with open(cg_dir + "tasks", "w", 0) as tasks, \
+                open(cg_dir + "devices.deny", "w", 0) as deny, \
+                open(cg_dir + "devices.allow", "w", 0) as allow:
+
+            # deny all by default
+            deny.write("a")
 
-    if len(sys.argv) != 2:
-        sys.exit("usage: {} device_node".format(sys.argv[0]))
+            # Only the "/dev/null" is required after CP-25630
+            whitelist = [
```

Comment:
Since there's only one element in this list you could probably collapse it all down to:

    allow.write(get_ctl("/dev/null", "rw"))

(Assuming get_ctl() is changed so that it can't fail.)

Reply:
done.